screenshot: fix %f corner case

WhitePeter opened the pull request:
URLs can end with trailing slashes (/) which in turn results in GNU basename[1] returning the empty string. Catch that case and fallback on the raw value of mpctx->filename. Subsequent path sanitation takes care of translating invalid path component chars.

Issue reproduction steps:
```
mpv \
    --screenshot-dir=$HOME/mpv-shots \
    --screenshot-template='%f/%P' \
    https://example.org/video/
```
This would result in %f expanding to '' and thus render screenshot-template an absolute path which, for some reason, would in turn take precedence of screenshot-dir and hence result in a non-writeable path, i.e. `/timestamp.ext`.

With this new approach the resulting path looks like this:
    `/home/user/mpv-shots/http:__example.org_video_/timestamp.ext`

Not particularly pretty but less ambiguous and more consistent over a diverse range of input paths. For transparency and reference also see [2] which sadly could not be revived.

[1] https://github.com/mpv-player/mpv/pull/14635#issuecomment-2269456187
[2] https://github.com/mpv-player/mpv/pull/17015

> Read this before you submit this pull request:
> https://github.com/mpv-player/mpv/blob/master/DOCS/contribute.md

Check!

guidocella commented:
According to DOCS/contribute.md GNU features should not be used and it may not work on Windows.

WhitePeter commented on player/screenshot.c:
```diff
@@ -300,9 +305,6 @@ static char *gen_fname(struct mp_cmd_ctx *cmd, const char *file_ext)
             dir = mp_get_user_path(t, ctx->mpctx->global, dir);
             fname = mp_path_join(NULL, dir, fname);
-
-            mp_mkdirp(dir);
```
As in a separate commit in this PR or another one?

guidocella commented on player/screenshot.c:
```diff
@@ -175,12 +175,17 @@ static char *create_fname(struct MPContext *mpctx, char *template,
+            if (!video_file) {
+                video_file = "NO_FILE";
+            } else {
                 video_file = mp_basename(mpctx->filename);
```
`filename` will contain the full URL, as https://github.com/mpv-player/mpv/pull/16932 is already changing. Huge URLs still have huge basenames since the long part is the query string.

guidocella replied:
Implemented in https://github.com/mpv-player/mpv/pull/17022

WhitePeter replied:
OK, I think I get it now. But I am not quite sure if this PR is the right place to do it.

Having said that, I do like the idea of having acutal consistency. Above special case only exists since other URLs are being treated differently because of that GNU basename quirk, owed to the non-mutating action promise, to just return `''` when there are trailing `/`. Thanks again for pointing it out, twice. ;)
Some special casing must exist somewhere already because otherwise, I would not see the full URL in `stats` or plain `${filename}`, in case of a trailing `/`. And there is `mp_is_url`.

But one possible caveat just occurred to me: `%F`, the supposed equivalent of `%{filename/no-ext}` can do funny things, e.g. if there is no `.` in the last path segment it will eat the top level domain. But somehow `%{filename/no-ext}` is immune to that.

WhitePeter replied:
Alright, I am sold. Please check out the new commits.

WhitePeter commented on player/screenshot.c:
```diff
@@ -70,7 +70,7 @@ void screenshot_init(struct MPContext *mpctx)
 static char *stripext(void *talloc_ctx, const char *s)
 {
-    const char *end = strrchr(s, '.');
+    const char *end = strrchr(mp_basename(s), '.');
```
@guidocella See, it *is* still useful. ;)

guidocella commented:
Yeah I also realized `mp_splitext` has those issue, and also see https://github.com/mpv-player/mpv/pull/17022#issuecomment-3507827060. We can continue in this PR.

WhitePeter commented:
> It's also worth noting that we also have `bstr_strip_ext` and `bstr_get_ext` but those don't check for URLs like `mp_splitext` so I'm not sure how we should handle all these similar functions?

I didn't know about them. I just saw `mp_splitext` used in the `filename/no-ext` property getter and figured to get rid of the duplicated effort. Maybe the case of handling URLs with those is covered by teaching them `mp_basename`?

But I am having a hard time wrapping my head around bstr usage. Can you give me or point me to a quick rundown, what purpose they server? I know from the bstring(3) manual that they are not necessarily `\0`-terminated. But that does not seem to explain why they might be useful in handling paths. Or is it, so one can have a long string and split on `\0` like `xargs --null`?

Ideally there should only be one of each `*{strip,split,get}ext`, I guess. `mp_splitext` can do both in one call, get the extension and split it of. So maybe `*_{get,set}_ext` can become shorthand wrappers around it?

WhitePeter commented:
> There is nothing to outsource. Getting the filename is trivial.

OK, but I was looking at that whole long `switch` `case` and especially `%{` which reinvents `mp.get`.

> And core functionality should work in builds with Lua disabled.

I think the core functionality can be boiled down to `screenshot-to-file` then? But I did not know that Lua was optional, I've figured it a hard dependency for a long time. Maybe it's time to make it official? There be benefits!

WhitePeter commented:
> > OK, but I was looking at that whole long `switch` `case` and especially `%{` which reinvents `mp.get`.
> 
> No. Retrieving a single property is different from expanding property strings. And Lua functions are built on C functions, not the opposite.

Yes, I know that. But why use the C functions with all that memory management overhead, i.e. dragging the talloc context along, and all that makes C difficult to deal with, when a simple `mp.get` will do? C is for performance critical code which this is not.
 
> > I think the core functionality can be boiled down to `screenshot-to-file` then? But I did not know that Lua was optional, I've figured it a hard dependency for a long time. Maybe it's time to make it official? There be benefits!
> 
> Core functionality like playback and taking screenshots works fine without Lua, so it doesn't need to be required.

I am totally with you on playback, of course, but only half with screenshots. One can get 99% of `screenshot` with an external screenshooter  - yes, I measured that. ;) Embedding playback context in the filename is very much what I'd call an extension. IIRC that's what the advertisement said, when mpv introduced the Lua interface: it's for extensions to the core functionality; `screenshot-to-file` being the glue between core functionality and higher level extensions in my (ideal) book.

But I don't want to argue. I am just trying to make a case for some code trimming to get a leaner meaner mpv core which really only provides media *playback* as best as can be done. Screenshots are media *creation* and as such not a show stopper if they did not exist at all as concept in mpv; there was, after all, a time before `screenshot`.

WhitePeter commented:
One last thing I just have to say to this because it does demonstrate, that 99.9% can be achieved with plain `screenshot-to-file` and a key binding with **Lua disabled**.

Adding this to input.conf:
```
# interestingly enough ${screenshot-dir} does not work here
s async screenshot-to-file "~/tmp/mpv-shots/${filename}/${playback-time}.${screenshot-format}"
```
emulates my initial example (`--screenshot-dir=~/tmp/mpv-shots --screenshot-template='%f/%P'`) almost perfectly, except I have not found a millisecond resolution property, but that's beside the point.

Now I shut up about the matter. I will go through the chore list and come back when I'm done. Thanks for all your insights.

guidocella commented:
`${playback-time/full}` has milliseconds. I guess the main advantage of `--screenshot-template` over that is `%n`.

guidocella commented:
Thinking it through, the way you just suggested is the easiest since most `%` format specifiers are pointless and require you to learn a second system on top of properties. I'm not sure why wm4 ever added them (in 12c44610ad), I guess at least `${time-pos/full}` (bd1d3eaacb) and `${clock}` (953d225368) didn't exist at the time.

If we were to implement it now, we could get rid of `screenshot`, `--screenshot-template`, `--screenshot-dir` and just use properties. Default to something like `screenshot-to-file ~/${filename/no-ext} ${time-pos/full}` so `%n` is not needed. That's a better default than `mpv-shot%n` anyway which is too generic, though `:` will be replaced by `_` on Windows. `${path/directory}` to replace `%x` and `${clock}` formatting sub-properties could be added if needed, and would also be useful elsewhere.

You would have to rebind all 4 default screenshot bindings to change path though, so maybe they are worth having for that, but at least the format specifiers are not needed.

It's probably not worth breaking all user configs and libmpv clients now though. But we could recommend using only the properties in the documentation and only keep format specifiers for backwards compatibility.

WhitePeter commented:
Sorry again, I just had to harden that macro, otherwise, if used in the wrong expression, operator precedence might have silently changed the semantics, e.g. `x = y && mp_basename_or_url(z)` would have expanded to `x = y && mp_is_url(z) ? ...`.

kasper93 commented on misc/path_utils.c:
```diff
@@ -85,15 +85,27 @@ void mp_path_strip_trailing_separator(char *path)
+    // Skip all leading dots, not just for "hidden" unix files, otherwise we
+    // end up splitting a part of the filename sans leading dot.
+    while (*bn == '.')
+        bn++;
```
`strspn`

kasper93 commented on misc/path_utils.c:
```diff
@@ -85,15 +85,27 @@ void mp_path_strip_trailing_separator(char *path)
     return (char *)split + 1;
 }
 
+char *mp_stripext(void *talloc_ctx, const char *s)
```
this should return const char* instead of casting.
